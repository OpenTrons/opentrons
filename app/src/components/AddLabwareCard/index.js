// @flow
import * as React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Card } from '@opentrons/components'

import { getConfig } from '../../config'
import {
  changeCustomLabwareDirectory,
  addCustomLabware,
  clearAddCustomLabwareFailure,
  getAddLabwareFailure,
} from '../../custom-labware'

import { ManagePath } from './ManagePath'
import { AddLabware } from './AddLabware'
import { PortaledAddLabwareFailureModal } from './AddLabwareFailureModal'

import type { Dispatch } from '../../types'

// TODO(mc, 2019-10-17): i18n
const LABWARE_MANAGEMENT = 'Labware Management'

export function AddLabwareCard() {
  const dispatch = useDispatch<Dispatch>()
  const config = useSelector(getConfig)
  const addFailure = useSelector(getAddLabwareFailure)
  const labwarePath = config.labware.directory
  const handleChangePath = () => dispatch(changeCustomLabwareDirectory())
  const handleAddLabware = () => dispatch(addCustomLabware())
  const showAddFailure = addFailure.file || addFailure.errorMessage !== null

  return (
    <Card title={LABWARE_MANAGEMENT}>
      <ManagePath path={labwarePath} onChangePath={handleChangePath} />
      <AddLabware onAddLabware={handleAddLabware} />
      {showAddFailure && (
        <PortaledAddLabwareFailureModal
          {...addFailure}
          directory={labwarePath}
          onCancel={() => dispatch(clearAddCustomLabwareFailure())}
          onOverwrite={file => dispatch(addCustomLabware(file))}
        />
      )}
    </Card>
  )
}
