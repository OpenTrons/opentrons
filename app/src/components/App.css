/* TODO(mc, 2018-02-09): fix lint errors and remove this ignore */
/* stylelint-disable font-family-no-missing-generic-family-keyword  */
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.wrapper {
  width: 1024px;
  height: 100vh;
  display: flex;
  flex-direction: row;
  font-family: 'sans';
  position: fixed;
}
