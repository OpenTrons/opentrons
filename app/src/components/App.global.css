@import '@opentrons/components';

* {
  box-sizing: border-box;
}

html,
body {
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
  color: var(--c-font-dark);
}

a {
  text-decoration: none;
}

.fixed {
  position: fixed;
}

.relative {
  position: relative;
}

.absolute {
  position: absolute;
}

.absolute__fill {
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
}

.inline {
  display: inline;
}

.block {
  display: block;
}

.inline_block {
  display: inline-block;
}

/*** TYPOGRAPHY ***/
@font-face {
  font-family: 'trons';
  src: url('../fonts/Geogtq-Rg-webfont.woff2');
  font-weight: normal;
  font-style: normal;
}

@font-face {
  font-family: 'sans';
  src: url('../fonts/OpenSans-Regular-webfont.woff');
  font-weight: normal;
  font-style: normal;
}

.sans {
  font-family: sans;
}

.trons {
  font-family: trons;
}

/*** BUTTONS ***/
button {
  border: none;
}

button:focus {
  outline: 0;
}

.btn {
  height: 2.5rem;
  line-height: 2.5rem;
  border-radius: 2px;
  font-family: 'sans';
  font-size: 14px;
  text-transform: uppercase;
  cursor: pointer;
  background-color: transparent;
}

.btn:active {
  outline: none;
  border: none;
}

.disabled {
  pointer-events: none;
  opacity: 0.5;
}

.btn_light {
  border: 1px solid silver;
  background-color: #fff;
  color: #666;
}

.btn_light:hover {
  background-color: #eee;
}

.btn_dark {
  background-color: #9b9b9b;
  color: white;
}

.btn_dark:hover {
  background-color: #777;
}

.btn_blue {
  background-color: #006fff;
  color: white;
}

.btn_blue:hover {
  background-color: #00f;
}
