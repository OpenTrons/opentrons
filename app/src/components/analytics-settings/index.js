// @flow
import AnalyticsSettingsCard from './AnalyticsSettingsCard'
import AnalyticsSettingsModal from './AnalyticsSettingsModal'

export { AnalyticsSettingsCard, AnalyticsSettingsModal }
