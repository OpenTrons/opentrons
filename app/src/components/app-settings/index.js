// @flow
// app settings related components
export * from './UpdateAppModal'
export * from './AppSoftwareSettingsCard'
export * from './AppAdvancedSettingsCard'
