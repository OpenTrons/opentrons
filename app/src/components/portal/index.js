// @flow
import * as React from 'react'
import ReactDom from 'react-dom'

type PortalLevel = 'page' | 'top'

type Props = {|
  children: React.Node,
  level: PortalLevel,
|}

type State = {|
  hasRoot: boolean,
|}

const PORTAL_ROOT_ID_BY_LEVEL: { [PortalLevel]: string } = {
  page: '__otAppModalPortalRoot',
  top: '__otAppTopPortalRoot',
}

const getPortalRoot = level =>
  global.document.getElementById(PORTAL_ROOT_ID_BY_LEVEL[level])

export function PortalRoot(): React.Node {
  return <div id={PORTAL_ROOT_ID_BY_LEVEL.page} />
}

export function TopPortalRoot(): React.Node {
  return <div id={PORTAL_ROOT_ID_BY_LEVEL.top} />
}

// the children of Portal are rendered into the PortalRoot if it exists in DOM
export class Portal extends React.Component<Props, State> {
  $root: ?Element

  static defaultProps: {| level: PortalLevel |} = {
    level: 'page',
  }

  constructor(props: Props) {
    super(props)
    this.$root = getPortalRoot(props.level)
    this.state = { hasRoot: !!this.$root }
  }

  // on first launch, $portalRoot isn't in DOM; double check once we're mounted
  // TODO(mc, 2018-10-08): prerender UI instead
  componentDidMount() {
    if (!this.$root) {
      this.$root = getPortalRoot(this.props.level)
      this.setState({ hasRoot: !!this.$root })
    }
  }

  render(): React.Portal | null {
    if (!this.$root) return null
    return ReactDom.createPortal(this.props.children, this.$root)
  }
}
