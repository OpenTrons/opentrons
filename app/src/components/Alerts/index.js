// @flow
import * as React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import head from 'lodash/head'

import * as AppAlerts from '../../alerts'
import { LostConnectionAlert } from '../LostConnectionAlert'
import { AnalyticsSettingsModal } from '../analytics-settings'
import { U2EDriverOutdatedAlert } from './U2EDriverOutdatedAlert'

import type { State, Dispatch } from '../../types'
import type { AlertId } from '../../alerts/types'

export function Alerts(): React.Node {
  const dispatch = useDispatch<Dispatch>()

  // TODO(mc, 2020-05-07): move head logic to selector with alert priorities
  const activeAlert: AlertId | null = useSelector((state: State) => {
    return head(AppAlerts.getActiveAlerts(state)) ?? null
  })

  const dismissAlert = remember => {
    if (activeAlert) {
      dispatch(AppAlerts.alertDismissed(activeAlert, remember))
    }
  }

  return (
    <>
      {/* TODO(mc, 2020-05-07): LostConnectionAlert currently controls its own
          render; move its logic into `state.alerts` */}
      <LostConnectionAlert />
      {/* TODO(mc, 2020-05-07): AnalyticsSettingsModal currently controls its
          own render; move its logic into `state.alerts` */}
      <AnalyticsSettingsModal />

      {activeAlert === AppAlerts.ALERT_U2E_DRIVER_OUTDATED ? (
        <U2EDriverOutdatedAlert dismissAlert={dismissAlert} />
      ) : null}
    </>
  )
}
