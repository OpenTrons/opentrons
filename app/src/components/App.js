import React from 'react'
import { Switch, Route, Redirect } from 'react-router'

import NavBar from './nav-bar'

import { PageWrapper } from '../components/Page'
import SidePanel from '../pages/SidePanel'
import Robots from '../pages/Robots'
import More from '../pages/More'
import Upload from '../pages/Upload'
import Calibrate from '../pages/Calibrate'
import Run from '../pages/Run'
import { PortalRoot as ModalPortalRoot } from './portal'

import './App.global.css'
import styles from './App.css'

export default function App() {
  return (
    <div className={styles.wrapper} onDragOver={stopEvent} onDrop={stopEvent}>
      <NavBar />
      <SidePanel />
      <PageWrapper>
        <Switch>
          <Redirect from="(.*)/index.html" to="/" />
          <Redirect exact from="/" to="/robots" />
          <Route path="/robots/:name?" component={Robots} />
          <Route path="/menu" component={More} />
          <Route path="/upload" component={Upload} />
          <Route path="/calibrate" component={Calibrate} />
          <Route path="/run" component={Run} />
        </Switch>
        <ModalPortalRoot />
      </PageWrapper>
    </div>
  )
}

function stopEvent(event) {
  event.preventDefault()
}
