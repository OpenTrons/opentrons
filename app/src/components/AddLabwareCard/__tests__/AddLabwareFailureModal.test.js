// @flow
import * as React from 'react'
import { shallow, mount } from 'enzyme'

import { AlertModal } from '@opentrons/components'
import { Portal } from '../../portal'
import {
  PortaledAddLabwareFailureModal,
  AddLabwareFailureModal,
} from '../AddLabwareFailureModal'

import * as LabwareFixtures from '../../../custom-labware/__fixtures__'

describe('AddLabwareFailureModal', () => {
  const mockDirectory = '/path/to/labware'
  const mockOnCancel = jest.fn()
  const mockOnOverwrite = jest.fn()
  const emptyProps = {
    directory: mockDirectory,
    file: null,
    errorMessage: null,
    onCancel: mockOnCancel,
    onOverwrite: mockOnOverwrite,
  }

  afterEach(() => {
    jest.resetAllMocks()
  })

  test('component renders', () => {
    const wrapper = shallow(<AddLabwareFailureModal {...emptyProps} />)

    expect(wrapper).toMatchSnapshot()
  })

  test('renders inside a Portal', () => {
    const wrapper = shallow(<PortaledAddLabwareFailureModal {...emptyProps} />)
    const portal = wrapper.find(Portal)
    const modal = portal.find(AddLabwareFailureModal)

    expect(modal.props()).toEqual(emptyProps)
  })

  test('renders an AlertModal', () => {
    const wrapper = shallow(<AddLabwareFailureModal {...emptyProps} />)

    expect(wrapper.exists(AlertModal)).toBe(true)
  })

  test('renders a cancel button that calls props.onCancel', () => {
    const wrapper = mount(<AddLabwareFailureModal {...emptyProps} />)
    const button = wrapper.findWhere(
      c => c.type() === 'button' && c.text().toLowerCase() === 'cancel'
    )

    button.invoke('onClick')()
    expect(mockOnCancel).toHaveBeenCalled()
  })

  test('renders proper title for error', () => {
    const wrapper = mount(
      <AddLabwareFailureModal {...emptyProps} errorMessage="AHHH!" />
    )
    const html = wrapper.html()

    expect(wrapper.find(AlertModal).prop('heading')).toEqual(
      'Unable to add labware'
    )
    expect(html).toMatch(/Unable to copy labware/)
    expect(html).toContain('AHHH!')
  })

  describe('invalid files', () => {
    const render = file => {
      return mount(<AddLabwareFailureModal {...emptyProps} file={file} />)
    }

    test('renders proper copy for invalid file', () => {
      const file = LabwareFixtures.mockInvalidLabware
      const wrapper = render(file)
      const html = wrapper.html()

      expect(wrapper.find(AlertModal).prop('heading')).toEqual(
        'Invalid labware definition'
      )
      expect(html).toMatch(/not a valid Opentrons labware definition/)
      expect(html).toContain(file.filename)
    })

    test('renders proper copy for an Opentrons conflicting file', () => {
      const file = LabwareFixtures.mockOpentronsLabware
      const wrapper = render(file)
      const html = wrapper.html()

      expect(wrapper.find(AlertModal).prop('heading')).toEqual(
        'Conflict with Opentrons labware'
      )
      expect(wrapper.html()).toMatch(
        /conflicts with an Opentrons standard definition/
      )
      expect(html).toContain(file.definition.metadata.displayName)
      expect(html).toContain(file.definition.parameters.loadName)
      expect(html).toContain(file.filename)
    })

    test('renders proper copy for an duplicate file', () => {
      const file = LabwareFixtures.mockDuplicateLabware
      const wrapper = render(file)
      const html = wrapper.html()

      expect(wrapper.find(AlertModal).prop('heading')).toEqual(
        'Overwrite duplicate labware?'
      )
      expect(html).toMatch(/already exists/)
      expect(html).toContain(file.definition.metadata.displayName)
      expect(html).toContain(file.definition.parameters.loadName)
      expect(html).toContain(file.filename)
    })

    test('duplicate file adds overwrite button', () => {
      const file = LabwareFixtures.mockDuplicateLabware
      const wrapper = render(file)
      const button = wrapper.findWhere(
        c => c.type() === 'button' && /overwrite/i.test(c.text())
      )

      button.invoke('onClick')()
      expect(mockOnOverwrite).toHaveBeenCalledWith(file)
    })
  })
})
