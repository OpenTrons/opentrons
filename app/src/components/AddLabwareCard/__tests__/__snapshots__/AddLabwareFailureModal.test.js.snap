// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLabwareFailureModal component renders 1`] = `
<AlertModal
  alertOverlay={true}
  buttons={
    Array [
      Object {
        "children": "cancel",
        "className": "button",
        "onClick": [MockFunction],
      },
    ]
  }
  heading="Unable to add labware"
>
  <p>
    Unable to copy labware file to
     
    <span
      className="code"
    >
      /path/to/labware
    </span>
    . 
    Please check that you are able to open this file and add files to your labware source folder, or contact Opentrons Support for help.
  </p>
</AlertModal>
`;
